pyteal/ast/router.py

```diff
@@ -0,0 +1,583 @@
+                        )
+                    if handler.subroutine.argument_count() != 0:
+                        raise TealInputError(
+                            f"abi-returning subroutine call should take 0 arg for bare-app call. "
+                            f"this abi-returning subroutine takes {handler.subroutine.argument_count()}."
+                        )
+                    return Seq(cast(Expr, handler()), Approve())
+                case _:
+                    raise TealInputError(
+                        "bare appcall can only accept: none type Expr, or Subroutine/ABIReturnSubroutine with none return and no arg"
+                    )
+        else:
+            if not isinstance(handler, ABIReturnSubroutine):
+                raise TealInputError(
+                    f"method call should be only registering ABIReturnSubroutine, got {type(handler)}."
+                )
+            if not handler.is_abi_routable():
+                raise TealInputError(
+                    f"method call ABIReturnSubroutine is not routable "
+                    f"got {handler.subroutine.argument_count()} args with {len(handler.subroutine.abi_args)} ABI args."
+                )
+
+            arg_type_specs = cast(
+                list[abi.TypeSpec], handler.subroutine.expected_arg_types
+            )
+            if handler.subroutine.argument_count() > METHOD_ARG_NUM_LIMIT:
+                last_arg_specs_grouped = arg_type_specs[METHOD_ARG_NUM_LIMIT - 1 :]
+                arg_type_specs = arg_type_specs[: METHOD_ARG_NUM_LIMIT - 1]
+                last_arg_spec = abi.TupleTypeSpec(*last_arg_specs_grouped)
+                arg_type_specs.append(last_arg_spec)
+
+            arg_abi_vars: list[abi.BaseType] = [
+                type_spec.new_instance() for type_spec in arg_type_specs
+            ]
+            decode_instructions: list[Expr] = [
+                arg_abi_vars[i].decode(Txn.application_args[i + 1])
+                for i in range(len(arg_type_specs))
+            ]
+
+            if handler.subroutine.argument_count() > METHOD_ARG_NUM_LIMIT:
+                tuple_arg_type_specs: list[abi.TypeSpec] = cast(
+                    list[abi.TypeSpec],
+                    handler.subroutine.expected_arg_types[METHOD_ARG_NUM_LIMIT - 1 :],
+                )
+                tuple_abi_args: list[abi.BaseType] = [
+                    t_arg_ts.new_instance() for t_arg_ts in tuple_arg_type_specs
+                ]
+                last_tuple_arg: abi.Tuple = cast(abi.Tuple, arg_abi_vars[-1])
+                de_tuple_instructions: list[Expr] = [
+                    last_tuple_arg[i].store_into(tuple_abi_args[i])
+                    for i in range(len(tuple_arg_type_specs))
+                ]
+                decode_instructions += de_tuple_instructions
+                arg_abi_vars = arg_abi_vars[:-1] + tuple_abi_args
```

Comment:
Why remove the last element from `arg_abi_vars`?

Reply:
`arg_abi_vars` are generated from `arg_type_specs`, while `arg_type_specs` has length at most 15.
The last type-spec or abi-var is an tuple before we de-tuple them into stack arguments, to feed them to subroutine.

Such "last (tupled) element from `arg_abi_vars`" should be discarded in this case, or the input to underlying subroutine would be wrong.